quantile/cm/stream.go

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cm
+
+import "math"
+
+const (
+	minSamplesToCompress = 3
+)
+
+// stream represents a data stream
+type stream struct {
+	capacity   int        // stream capacity
+	eps        float64    // desired epsilon for errors
+	quantiles  []float64  // sorted target quantiles
+	samplePool SamplePool // pool of samples
+	floatsPool FloatsPool // pool of float64 slices
+
+	closed          bool       // whether the stream is closed
+	numValues       int64      // number of values
+	bufLess         minHeap    // sample buffer whose value is less than that at the insertion cursor
+	bufMore         minHeap    // sample buffer whose value is more than that at the insertion cursor
+	samples         sampleList // sample list
+	insertCursor    *Sample    // insertion cursor
+	compressCursor  *Sample    // compression cursor
+	compressMinRank int64      // compression min rank
+}
+
+// NewStream creates a new sample stream
+// TODO(xichen): the stream object itself should be pooled too
+// TODO(xichen): add metrics
+func NewStream(capacity int, opts Options) Stream {
+	samplePool := opts.SamplePool()
+	floatsPool := opts.FloatsPool()
```

Comment:
I called it `floatsPool` to follow the naming convention for slice pools (e.g.,`BytesPool`). Can call it `floatSlicePool` to make it more explicit